### PR TITLE
docs: proposal for compat mode

### PR DESCRIPTION
**Pull request by brandonjpierce**
This is based on a conversation with Chris and Xiaoji from the Deckgl team. The idea is to support the current official Deckgl React bindings with a backwards compatible API.

Primary content is in `proposal.md` and `design.md`

### CODE REVIEW

File: openspec/changes/deckgl-react-compat/proposal.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+Users of the official `@deck.gl/react` bindings can't migrate to our fiber renderer without rewriting their code. A compatibility layer lets them switch by changing one import, then migrate to the native API gradually.
+
+## What Changes
+
+- Add `/compat` deep import path to `@deckgl-fiber-renderer/dom` package exporting backwards-compatible API
+- Create `DeckGL` component wrapper that adapts official prop API to our native `Deckgl` component
+- Implement `DeckGLContext` that exposes viewport, deck instance, and eventManager to child components
+- Export JSX layer wrapper components (ScatterplotLayer, ArcLayer, etc.) that bridge React elements to our reconciler's `<layer>` primitive
+- Export JSX view wrapper components (MapView, FirstPersonView, etc.) using our reconciler's `<view>` primitive
+- Implement `useWidget` hook for deck.gl widget integration (no pre-built widget components)
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/proposal.md]: Does fiber.gl have a `<widget>`? The official react binding for widgets doesn't add anything over using the normal `widgets` props, so I'd prefer to drop the wrappers.

File: openspec/changes/deckgl-react-compat/proposal.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+_None - this change only adds new capabilities without modifying existing behavior_
+
+## Impact
+
+**Affected Packages:**
+- `@deckgl-fiber-renderer/dom`: Add `/compat` directory and package.json export entry
+
+**New Dependencies:**
+- `@deck.gl/widgets` (peer dependency for useWidget hook)
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/proposal.md]: Connected to above, I'd prefer to not use the `useWidget` hook.

File: openspec/changes/deckgl-react-compat/design.md [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+- Our `@deckgl-fiber-renderer/dom` package exports `Deckgl` component with JSX layer syntax: `<Deckgl><layer layer={new ScatterplotLayer({...})} /></Deckgl>`
+- Official API uses `<DeckGL layers={[new ScatterplotLayer({...})]} />` or JSX layers: `<DeckGL><ScatterplotLayer {...} /></DeckGL>`
+- No migration path exists - users must rewrite all deck.gl code to switch
+
+**Constraints:**
+- Must not modify existing native API (non-breaking change)
+- Must work with our reconciler's persistence mode architecture
+- Cannot depend on `@deck.gl/react` package (creates circular dependency conceptually)
+- Must maintain layer ID requirements (critical for deck.gl diffing)
+- Limited to what our current reconciler supports (no positioned render callbacks)
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: What does this mean?

File: openspec/changes/deckgl-react-compat/design.md [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+- Current `@deck.gl/react` users wanting to try our renderer
+- Teams with existing deck.gl codebases
+- New users familiar with official API patterns
+
+## Goals / Non-Goals
+
+**Goals:**
+- Users can import from `/compat` instead of `@deck.gl/react` with minimal code changes
+- JSX layer syntax works: `<DeckGL><ScatterplotLayer /></DeckGL>`
+- JSX view syntax works: `<DeckGL><MapView /></DeckGL>`
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]:
```suggestion
- JSX view syntax works: `<DeckGL><MapView /></DeckGL>`
- JSX widget syntax works: `<DeckGL><ZoomWidget /></DeckGL>`
```

File: openspec/changes/deckgl-react-compat/design.md [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+- Imperative ref methods match official API
+- Context provides viewport/deck access for child components
+- Clear documentation of differences and migration path
+
+**Non-Goals:**
+- 100% behavioral compatibility (internal implementation can differ)
+- Positioned render callbacks: `{({viewport}) => <div>...</div>}`
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: Ah, I didn't know this was called a positioned render callback

File: openspec/changes/deckgl-react-compat/design.md [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+- Pre-built widget components (only provide hook)
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]:
```suggestion
```

File: openspec/changes/deckgl-react-compat/design.md [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+  "exports": {
+    ".": "./dist/index.js",
+    "./compat": "./dist/compat/index.js",
+    "./components": "./dist/components.js",
+    "./hooks": "./dist/hooks.js"
+  }
+}
+```
+
+### Decision 2: JSX Layer Strategy - Thin Wrapper Components
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: See comment

File: openspec/changes/deckgl-react-compat/design.md [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+- 100% behavioral compatibility (internal implementation can differ)
+- Positioned render callbacks: `{({viewport}) => <div>...</div>}`
+- Pre-built widget components (only provide hook)
+- Custom Deck class injection: `<DeckGL Deck={CustomDeck} />`
+- Supporting `_customRender` prop
+- Direct `canvas`/`gl` props (use `interleaved` instead)
+
+## Decisions
+
+### Decision 1: Package Structure - Deep Import Path
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: Agree

File: openspec/changes/deckgl-react-compat/design.md [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+  const { children, ...layerProps } = props;
+  return <layer layer={new _ScatterplotLayer(layerProps)}>{children}</layer>;
+}
+```
+
+**Alternatives Considered:**
+- **Extract JSX then bypass reconciler:** Matches official behavior more closely but loses reconciler benefits (efficient updates, lifecycle hooks, suspense support)
+- **Dynamic proxy pattern:** Zero maintenance but less type-safe, harder to tree-shake, less explicit
+
+**Layer Coverage:**
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: It'll be important to have full coverable for a non-breaking migration. Perhaps we need more exports that mirror the official deck layer modules? e.g.

`@deckgl-fiber-renderer/compat/layers`
`@deckgl-fiber-renderer/compat/geo-layers`
`@deckgl-fiber-renderer/compat/aggregation-layers`
`@deckgl-fiber-renderer/compat/mesh-layers`

Not every user needs all of them, but all are wrapped for an easy migration

File: openspec/changes/deckgl-react-compat/design.md [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+- Start with 15-20 most common layers manually
+- Core layers (`@deck.gl/layers`): ScatterplotLayer, ArcLayer, LineLayer, GeoJsonLayer, PolygonLayer, PathLayer, IconLayer, TextLayer, ColumnLayer, GridCellLayer, PointCloudLayer, BitmapLayer, SolidPolygonLayer
+- Geo layers (`@deck.gl/geo-layers`): H3HexagonLayer, S2Layer, GreatCircleLayer, TileLayer, MVTLayer
+- Can add codegen script later if users need more layers
+
+### Decision 3: DeckGL Component Architecture - Thin Adapter
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: I'm assuming Deckgl is the component this library ships?

File: openspec/changes/deckgl-react-compat/design.md [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+    </CompatContextProvider>
+  );
+});
+```
+
+**Alternatives Considered:**
+- **Fork official implementation:** Higher compatibility but massive code duplication and maintenance burden
+- **Hybrid with edge case handling:** More complex, unclear when to use which path
+
+### Decision 4: JSX Element Handling - Let Reconciler Do The Work
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: Agree

File: openspec/changes/deckgl-react-compat/design.md [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+
+  const Provider = ContextProvider || DeckGLContext.Provider;
+  
+  return <Provider value={contextValue}>{children}</Provider>;
+}
+```
+
+**Limitations Accepted:**
+- Context updates only when deck instance changes (may be less granular than official)
+- May not expose all internal deck.gl properties
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: Which are excluded?

File: openspec/changes/deckgl-react-compat/design.md [hunk 13]
```diff
@@ -0,0 +1,511 @@
...
+- Documented in migration guide
+
+### Decision 6: Widget Support - Hook Only
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: Disagree, I'd prefer to drop the official react widget code entirely (it's buggy and the widgets exported by the widget module should be each for the reconciler to support since they're natively declarative and have a pure JS interface)

File: openspec/changes/deckgl-react-compat/design.md [hunk 14]
```diff
@@ -0,0 +1,511 @@
...
+
+**Alternatives Considered:**
+- **Pre-wrapped common widgets:** Better DX but 5-10KB added, maintenance burden
+- **All widgets:** Full compatibility but 15-20KB added, high maintenance
+- **Codegen widgets:** Automated but build complexity
+
+**Dependencies:**
+- `@deck.gl/widgets` as peer dependency (user installs)
+
+### Decision 7: Type Definitions - Create Our Own
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: This is probably fine. Better to avoid circular issues

File: openspec/changes/deckgl-react-compat/design.md [hunk 15]
```diff
@@ -0,0 +1,511 @@
...
+export type LayerProps<T> = T & {
+  children?: ReactNode;
+};
+```
+
+**Alternatives Considered:**
+- **Re-export from `@deck.gl/react`:** Always in sync but requires dependency
+- **Hybrid approach:** Partial re-export plus custom types, split maintenance
+
+### Decision 8: Error Handling - Development Warnings
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: I assume this is only for select compat logs? In which case, I agree.

File: openspec/changes/deckgl-react-compat/design.md [hunk 16]
```diff
@@ -0,0 +1,511 @@
...
+
+### [Risk] JSX layer re-render behavior differs from official
+**Impact:** Layers may re-initialize at different times than official bindings
+
+**Why it happens:** Our reconciler uses persistence mode (creates new instances on every change), official bindings use mutation mode
+
+**Mitigation:**
+- Stable layer IDs prevent expensive re-initialization (our reconciler validates this)
+- Document the difference in migration guide
+- In practice, deck.gl handles both patterns well due to efficient ID-based diffing
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: Agreed, this should be a non-issue

File: openspec/changes/deckgl-react-compat/design.md [hunk 17]
```diff
@@ -0,0 +1,511 @@
...
+**Impact:** Some child components reading context may break
+
+**Why it happens:** We map from our store to official context shape, may miss edge cases
+
+**Mitigation:**
+- Document which context properties are available
+- Test with common child component patterns
+- Users can fall back to our native hooks (`useDeckgl`)
+
+### [Risk] Missing layer/view wrappers
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: We should wrap all official exports

File: openspec/changes/deckgl-react-compat/design.md [hunk 18]
```diff
@@ -0,0 +1,511 @@
...
+**Impact:** More work than official pre-built widget components
+
+**Why it happens:** Trade-off for smaller bundle and less maintenance
+
+**Mitigation:**
+- Provide cookbook with examples of wrapping common widgets
+- Pattern is simple (5 lines per widget)
+- Most users only need 1-3 widgets
+
+### [Risk] Type definitions drift from official API
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: We can live with this risk. It doesn't change often and we'll deprecate the official react bindings over time.

File: openspec/changes/deckgl-react-compat/design.md [hunk 19]
```diff
@@ -0,0 +1,511 @@
...
+**Impact:** Types may become inaccurate as deck.gl evolves
+
+**Why it happens:** We maintain our own types instead of depending on `@deck.gl/react`
+
+**Mitigation:**
+- Types based on stable `@deck.gl/core` which changes less frequently
+- Test types with real deck.gl usage
+- Users report type issues via GitHub
+
+### [Risk] Performance overhead from wrapper layer
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: I imagine this is typically negligible?

File: openspec/changes/deckgl-react-compat/design.md [hunk 20]
```diff
@@ -0,0 +1,511 @@
...
+- Type tests for IntelliSense compatibility
+- Manual testing with example app migrated from official API
+
+## Open Questions
+
+### Q1: Should we support more than 15-20 layers initially?
+**Options:**
+- **A:** Start with 15-20 core + geo layers (recommended)
+- **B:** Add mesh + aggregation layers too (~30 total)
+- **C:** Codegen all layers from start (~40 total)
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: Option C

File: openspec/changes/deckgl-react-compat/design.md [hunk 21]
```diff
@@ -0,0 +1,511 @@
...
+**Options:**
+- **A:** Hook only (current plan)
+- **B:** Add 3-5 most common widgets (Zoom, Compass, Fullscreen)
+- **C:** Add all widgets
+
+**Recommendation:** Option A initially. Revisit if many users request it.
+
+### Q3: Error handling - warnings or errors for unsupported features?
+**Options:**
+- **A:** Warnings only (current plan)
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: Option A seems fine

File: openspec/changes/deckgl-react-compat/design.md [hunk 22]
```diff
@@ -0,0 +1,511 @@
...
+
+**Recommendation:** Option A. Profile if users report performance issues.
+
+### Q5: Documentation format?
+**Options:**
+- **A:** Single migration guide page
+- **B:** Separate pages for each feature area
+- **C:** Interactive migration tool
+
+**Recommendation:** Option B. Easier to navigate, better for SEO.
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: Option B sounds fine.

File: openspec/changes/deckgl-react-compat/design.md [hunk 23]
```diff
@@ -0,0 +1,511 @@
...
+
+### Q3: Error handling - warnings or errors for unsupported features?
+**Options:**
+- **A:** Warnings only (current plan)
+- **B:** Errors in development, warnings in production
+- **C:** Optional strict mode
+
+**Recommendation:** Option A. Can add strict mode later.
+
+### Q4: Should context update more granularly?
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: I'd need more context to have an opinion

File: openspec/changes/deckgl-react-compat/design.md [hunk 24]
```diff
@@ -0,0 +1,511 @@
...
+  }, [deckgl]);
+
+  const Provider = ContextProvider || DeckGLContext.Provider;
+  
+  return <Provider value={contextValue}>{children}</Provider>;
+}
+```
+
+**Limitations Accepted:**
+- Context updates only when deck instance changes (may be less granular than official)
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/design.md]: I'm not sure what the implications are of this decision

**Review by chrisgervang — COMMENTED** [review]
Thank you! Here is my long overdue review

File: openspec/changes/deckgl-react-compat/proposal.md (hunk 1 above)
Comment (brandonjpierce) [reply openspec/changes/deckgl-react-compat/proposal.md]: Not presently, I am in agreement with you that providing a `<widget />` primitive would not net us any benefits so I am excluding it from the updated proposal (pushing soon)

File: openspec/changes/deckgl-react-compat/design.md (hunk 3 above)
Comment (brandonjpierce) [reply openspec/changes/deckgl-react-compat/design.md]: LLM's using weird phrasing 😅. Basically it means we don't support children as a function pattern (by design) e.g.

```
<Parent>
  {() => <Child />}
</Parent>
```

File: openspec/changes/deckgl-react-compat/design.md (hunk 9 above)
Comment (brandonjpierce) [reply openspec/changes/deckgl-react-compat/design.md]: Yeah we could certainly do that with `package.json#exports` field, i'll add that into the newest proposal

File: openspec/changes/deckgl-react-compat/design.md (hunk 13 above)
Comment (brandonjpierce) [reply openspec/changes/deckgl-react-compat/design.md]: Agreed, dropping.

File: openspec/changes/deckgl-react-compat/proposal.md (hunk 1 above)
Comment (chrisgervang) [reply openspec/changes/deckgl-react-compat/proposal.md]: Oh sorry, I think there may be a misunderstanding - I mean to say deck widgets would ideally be treated the same as layer and view components - the reconciler passes them through to deck, and reconciles which view they are nested under.

While widgets are UI elements, they are 100% managed internally and don't need any react hooks or _react_ wrappers.

Comment (chrisgervang) [reply openspec/changes/deckgl-react-compat/proposal.md]: My recommendation is to include a `<widget>` fiber JSX primitive, but to not use typical react UI hooks, components, etc

Comment (brandonjpierce) [reply openspec/changes/deckgl-react-compat/proposal.md]: Gotcha, I can include a secondary proposal isolated to just the widget change

**Comment by brandonjpierce** [comment]
@chrisgervang made a large update to the `design.md` and `proposal.md` documents (apologies for the churn). Largely just took your feedback and explicitly worked it into the design. Working on getting the widget proposal produced now. 

Going forward the only files you need to review are:

- openspec/changes/deckgl-react-compat/proposal.md
- openspec/changes/deckgl-react-compat/design.md

File: openspec/changes/deckgl-react-compat/pr-questions.md [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+   - deferred to future change
+   - intentionally unsupported
+9. If JSX widget syntax is deferred, should this change mention the future direction at all, or omit it entirely to avoid implied commitment?
+10. Does the existing change still deserve the same slug/scope (`deckgl-react-compat`), or has the PR feedback narrowed it to a more specific compatibility slice?
+
+## Deferred follow-up topic (explicitly not in scope here)
+
+Potential future change:
+- Widget primitive / JSX widget compatibility, possibly via:
+  - `<widget widget={new Widget()} />`
```
Comment (chrisgervang) [inline openspec/changes/deckgl-react-compat/pr-questions.md:126]: That interface should work for widget. No hooks needed, since widgets are pure JS.

**Review by chrisgervang — APPROVED** [review]
